File: src/main/java/org/micoli/php/service/filesystem/FileListener.java
package org.micoli.php.service.filesystem;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.newvfs.BulkFileListener;
import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.jetbrains.annotations.NotNull;

public class FileListener<Id> {
    protected static final Logger LOGGER = Logger.getInstance(FileListener.class.getSimpleName());
    boolean enabled;

    public interface VfsHandler<T> {
        void vfsHandle(T id, VirtualFile file);
    }

    private final BulkFileListener vfsListener;
    Map<Id, List<PathMatcher>> patterns = new HashMap<>();

    public FileListener(VfsHandler<Id> handler) {
        this.vfsListener = new BulkFileListener() {
            @Override
            public void after(@NotNull List<? extends VFileEvent> events) {
                events.forEach(event -> {
                    if (!enabled) {
                        return;
                    }
                    VirtualFile file = event.getFile();
                    if (file == null || file.isDirectory()) {
                        return;
                    }
                    Path path = Paths.get(file.getPath());
                    patterns.forEach((id, pathMatchers) -> {
                        pathMatchers.forEach(pathMatcher -> {
                            if (pathMatcher.matches(path)) {
                                handler.vfsHandle(id, file);
                            }
                        });
                    });
                });
            }
        };
    }

    public void setPatterns(Map<Id, List<PathMatcher>> patterns) {
        this.enabled = true;
        this.patterns = patterns;
    }

    public void reset() {
        this.enabled = false;
        this.patterns = new HashMap<>();
    }

    public boolean isEnabled() {
        return enabled;
    }

    public Map<Id, List<PathMatcher>> getPatterns() {
        return patterns;
    }

    public BulkFileListener getVfsListener() {
        return this.vfsListener;
    }
}

File: src/main/kotlin/org/micoli/php/service/filesystem/FileListener.kt
package org.micoli.php.service.filesystem

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.vfs.newvfs.BulkFileListener
import com.intellij.openapi.vfs.newvfs.events.VFileEvent
import com.intellij.util.messages.MessageBus
import java.nio.file.Paths

class FileListener<Id>(handler: VfsHandler<Id>, messageBus: MessageBus?) {
    var isEnabled: Boolean = false

    interface VfsHandler<T> {
        fun vfsHandle(id: T, file: VirtualFile)
    }

    val vfsListener: BulkFileListener
    private var patterns: MutableMap<Id, Watchee> = HashMap()

    init {
        this.vfsListener =
            object : BulkFileListener {
                override fun after(events: MutableList<out VFileEvent>) {
                    if (!isEnabled) {
                        return
                    }
                    events.forEach { event: VFileEvent ->
                        val file = event.file ?: return@forEach
                        val path = Paths.get(file.path)
                        patterns.forEach { (id: Id, watchee: Watchee) ->
                            if (watchee.match(path, event)) {
                                handler.vfsHandle(id, file)
                            }
                        }
                    }
                }
            }
        subscribe(messageBus)
    }

    private fun subscribe(messageBus: MessageBus?) {
        if (messageBus == null) return
        messageBus
            .connect()
            .subscribe<BulkFileListener>(VirtualFileManager.VFS_CHANGES, vfsListener)
    }

    fun setPatterns(patterns: MutableMap<Id, Watchee>) {
        this.isEnabled = true
        this.patterns = patterns
    }

    fun getPatterns(): MutableMap<Id, Watchee> {
        return this.patterns
    }

    fun reset() {
        this.isEnabled = false
        this.patterns = HashMap()
    }

    companion object {
        val LOGGER: Logger = Logger.getInstance(FileListener::class.java.getSimpleName())
    }
}
